wearerequired/traduttore
# The Big 3.0 Update 💥

This is a first pass at making Traduttore way more flexible and working for (almost) all sorts of source code repositories.

I got a bit carried away during implementation, so this contains a lot of changes—but I think they're worth it.

Here's a breakdown:

* Uses Traduttore Registry to make the plugin translatable using our own translation platform running Traduttore (inception!).  (#73).  
  Right now only the plugin meta data and the Slack event names are translatable. In a next step we can extend this to the actual Slack notification messages and perhaps more.
* Adds source code loaders for Mercurial and Subversion repositories.  
  This means Mercurial repositories on Bitbucket can be properly downloaded.  
  While no Subversion repository provider (e.g. SourceForge) is directly supported right now, the loader could be used by developers with their own repository or developers who prefer using GitHub with Subversion. This can be achieved by hooking into the various filters to override the various class instances.
* Adds support for incoming webhooks from Bitbucket, GitHub, and GitLab (GitLab.com and self-managed)  
  This is by far the biggest change. It should especially allow using self-managed GitLab repositories without necessarily having to write any custom code to override URLs or things like that.  
  Upon handling a webhook, repository information like name and URLs are stored as meta data in the database. This means a self-managed GitLab repository will point to the right host and everything after the first incoming webhook.
* When scheduling code updates, it first unschedules any existing events.
* Lots of documentation changes.  
  This is far from complete, but is a good start.

Things I still need to do (any help appreciated):

* Manual Testing  
  This includes testing with actual Bitbucket and GitLab repositories.
* Documentation improvements  
  Document and verify steps for repository setup, as well as adding docs for all new filters.
* Getting the Loader tests to work  
  We don't want to test if Git/Mercurial/Subversion are actually working, but whether the behavior of the classes is correct. Otherwise tests fail if the software isn't installed on your machine (or the CI provider).  
  I've looked into https://github.com/php-mock/php-mock-phpunit and https://github.com/mikehaertl/php-shellcommand for mocking `exec()` calls and some sort of Command dependency injection.
* Analyze existing class logic  
  Might need to move some stuff around. For example `delete_local_repository()` might make more sense in the `Updater` instead of `Runner`.

Fixes #64, #65, #66, #73.
